core/src/token_info.rs

```diff
@@ -1,17 +1,21 @@
 use anyhow::Result;
+use futures::future::BoxFuture;
 use lazy_static::lazy_static;
+#[cfg(test)]
+use mockall::automock;
 use serde::Deserialize;
 use std::collections::HashMap;
 
 use crate::models::{TokenId, TokenInfo};
 pub mod hardcoded;
 
+#[cfg_attr(test, automock)]
 pub trait TokenInfoFetching: Send + Sync {
     /// Retrieves some token information from a token ID.
-    fn get_token_info(&self, id: TokenId) -> Result<TokenBaseInfo>;
+    fn get_token_info<'a>(&self, id: TokenId) -> BoxFuture<'a, Result<TokenBaseInfo>>;
 
     /// Returns a vector with all the token IDs available
-    fn all_ids(&self) -> Result<Vec<TokenId>>;
+    fn all_ids<'a>(&self) -> BoxFuture<'a, Result<Vec<TokenId>>>;
```

Comment:
Could you tell me a bit about the reason for introducing this... lifetime parameter? `<`a>`

Reply:
Great question!

So a BoxFuture always needs a lifetime (https://docs.rs/futures/0.3.5/futures/future/type.BoxFuture.html) which cannot be inferred from the [elision rules](https://doc.rust-lang.org/reference/lifetime-elision.html) therefore it has to be explicitly stated. 
Now, just using `'a` here might be pointless since no other reference on the method definition has the same lifetime (so we could use `'static` as well). If I understand correctly this lifetime basically tells the compiler what references it can borrow inside the future (since the future might live/be active beyond this function call).

In other cases we have shared the `'a` lifetime with `&'a self` which means that the future can borrow variables from self (we are basically guaranteeing the future will be consumed before self goes out of scope).

I didn't need to borrow `self` in the code so far, but maybe we should still allow it in the abstract trait? What do our rust experts think?

Reply:
> What do our rust experts think?

I think it was a great explanation!

> I didn't need to borrow self in the code so far, but maybe we should still allow it in the abstract trait? 

I think so, or else you will get "does not live long enough" errors if you borrow self for the future. When implementing the caching layer as you had suggested in the previous PR, this might become a problem as you may need to update cached values which would require `self` to at least live for part of the future.

Reply:
```rust
fn all_ids<'a>(&self) -> BoxFuture<'a, Result<Vec<TokenId>>>;
```

I know it's already changed but want to point out that I think this doesn't make sense. This signature says that the caller of the function can pick the lifetime of the return value freely because no other parameter uses this lifetime. This can't be right because the caller's choice doesn't actually influence body of this function. In practice it would cause all implementation to have to return `'static` because that is the only way you can guarantee to outlive `'a` because `'a` might also be `'static` in which case we should just write `'static` in the signature.